### ISSUE TITLE
Curb integration: Curl::Multi support

### ISSUE THREAD

**Issue by sriedel** [bblimke/webmock#535]
I have a project where I need to use Curl::Multi, and would love to unit test the methods. I could have a go at implementing the missing parts for the integration. I've had a look at lib/webmock/http_lib_adapters/curb_adapter.rb but don't really see whats missing for webmock to work with Curl::Multi. If you could give me any pointers on where to start, I can have a go at the integration.

**Comment by bblimke** [comment]
@sriedel Thanks for your offer of adding Curb::Multi adapter.

The problem with Curb::Multi was that lots of it's code was tightly coupled with C extensions,
and I could not find a way to get around it.

@phiggins also took and attempt to do that and ended up with the same problem. Please have a look at this old issue https://github.com/bblimke/webmock/issues/67

Perhaps something has changed in Curb::Multi implementation over the last 5 years and it's easier to mock it nowadays. @taf2 ?

**Comment by taf2** [comment]
@bblimke still pretty similar to 5 years ago... maybe the newisher ruby smaller ruby API I exposed in 2011 would help?  e.g.  Curl.get  Curl.post - it's pure ruby implementation and uses Curl::Multi under the covers.

**Comment by bblimke** [comment]
Thanks @taf2 

The idea is to mock the stuff "under the covers" in WebMock so people can use any interface they wish.

As a workaround we could just stub `Curl.get` and `Curl.post` if documented properly in WebMock docs. 
@sriedel would that work for you?

**Comment by sriedel** [comment]
I'll have a look. I have some vacation coming up, so I'll have some more time to look more deeply into things.
From what I'm seeing in the other http adapter integrations, basically I'll have to interact with the Webmock Callback and Request registries at the appropriate places, is that correct?

**Comment by bblimke** [comment]
@sriedel yes. Curb adapter is quite complex as an example, but you can have a look at some other simpler adapters first.

**Comment by sriedel** [comment]
@bblimke: I finally got around to looking around the adapters to get a feel for what is needed for Curl::Multi. A quick question regarding ruby backward compatibility: since 1.9.3 is already past it's end of life, are you still planning on supporting ruby down to 1.8.7, or are you planning to drop support for everything below 2.0 (for example) in the foreseeable future? Just checking to see which methods I can use :)

**Comment by bblimke** [comment]
@sriedel since it hasn't caused any harm or slowed down development, I was keeping WebMock compatible with 1.8.7, but perhaps it's time to move on.

@myronmarston any thoughts? do you want to keep vcr compatible with 1.8.7?

**Comment by sriedel** [comment]
FWIW, when I tried to follow the "build a webmock development environment" instructions in the README, I could not build jeweler any more with ruby 1.8.6 (rake needs at least 1.8.7) and Addressable warns about 1.8.7 being deprecated.

**Comment by sriedel** [comment]
I got an initial version working, however I made the mocked Curl::Multi class a wrapper around the mocked Curl::Easy class. However, within curb Curl::Easy is a wrapper around Curl::Multi. I feel I should try to straighten things out a bit in the code before handing it to you for review. Depending on other workloads, this may take a few more days.

**Comment by myronmarston** [comment]
I'm not longer maintaining VCR.  @krainboltgreene maintains it now.  I know he just released VCR 3.0 so he may have dropped support for 1.8.7 already, but you'll have to ask him.

**Comment by sriedel** [comment]
@bblimke: I just found an issue in the Curl::Easy integration: Curbs on_redirect callback isn't being handled. Would you prefer to have a separate branch for the fix, or should I do it while working on the Curl::Multi integration?

**Comment by bblimke** [comment]
@sriedel great! please submit a separate pull request.

**Comment by sriedel** [comment]
@bblimke: Could you have a look at https://github.com/bblimke/webmock/compare/master...sriedel:sr/curl_multi?expand=1 to see if this is acceptable to you, or if you'd like to have something changed/added/removed?

After trying to produce a cleaner version of the Curb webmock code by having Curl::Easy wrap around Curl::Multi and failing, I kept the version where Curl::Multi uses the mocked Curl::Easy calls internally. That way I get a lot of the features for free (e.g. deciding per call if a mocked or real request should be generated). And since Curb has some write-only options, I don't really see a way to cleanly change the code to reverse the wrapping logic.

Note: I know I'm still using the old-style rspec syntax. I'll change that to the new expect-type syntax before creating the pull request.

**Comment by bblimke** [comment]
@sriedel I appreciate you came up with a workaround! I'm slightly concerned about that approach though. You are basically ignoring Curb::Multi internals.

The fact that tests/specs pass with this wrapper, doesn't prove the code works with Curb::Multi,
as you have reimplemented Curb::Multi.

Your `WebMockCurlMulti` is a wrapper around Curl::Easy, but it could have been wrapper around `Net::HTTP` as well. Curb::Multi code is completely ignored, just the interface is there. Almost any other adapter could have been implemented by replacing original class with Net::HTTP wrapper, but it's not the point. We want to make sure we test that our code works against the original object.

I guess if we added any specs related to Curl::Multi behaviour only, which is no available in Curl::Easy,
these specs would fail as Curl::Easy doesn't match Curl::Multi behaviour.

I don't have any other solution though, that why there was no Curl::Multi adapter so far.

I don't want WebMock to give people false sense of that their code is well tested, while it's not really,
and I'm not sure this kind of approach should be part of WebMock.
I'm wondering if a separate project i.e. webmock-curbmulti wouldn't be a better approach,
making it explicit that when people require it, they overwrite Curb::Multi with a Curl::Easy wrapper.

Any thoughts?

**Comment by taf2** [comment]
I'm wondering if we might take a similar approach to the curb test suite
and actually have a local server running that it can connect too or use a
file URL

I'm sorry I have not had more time to fully understand what webmock is
providing to understand why curb is not easier to work with.

On Friday, November 13, 2015, Bartosz Blimke <notifications@github.com>
wrote:

> @sriedel <https://github.com/sriedel> I appreciate you came up with a
> workaround! I'm slightly concerned about that approach though. You are
> basically ignoring Curb::Multi internals.
>
> The fact that tests/specs pass with this wrapper, doesn't prove the code
> works with Curb::Multi,
> as you have reimplemented Curb::Multi.
>
> Your WebMockCurlMulti is a wrapper around Curl::Easy, but it could have
> been wrapper around Net::HTTP as well. Curb::Multi code is completely
> ignored, just the interface is there. Almost any other adapter could have
> been implemented by replacing original class with Net::HTTP wrapper, but
> it's not the point. We want to make sure we test that our code works
> against the original object.
>
> I guess if we added any specs related to Curl::Multi behaviour only, which
> is no available in Curl::Easy,
> these specs would fail as Curl::Easy doesn't match Curl::Multi behaviour.
>
> I don't have any other solution though, that why there was no Curl::Multi
> adapter so far.
>
> I don't want WebMock to give people false sense of that their code is well
> tested, while it's not really,
> and I'm not sure this kind of approach should be part of WebMock.
> I'm wondering if a separate project i.e. webmock-curbmulti wouldn't be a
> better approach,
> making it explicit that when people require it, they overwrite Curb::Multi
> with a Curl::Easy wrapper.
>
> Any thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bblimke/webmock/issues/535#issuecomment-156467798>.
>


-- 
*Todd Fisher*
Tel: +1 (410) 394-9424 | Toll Free: +1 (800) 577-1872
todd@calltrackingmetrics.com | calltrackingmetrics.com

**Comment by sriedel** [comment]
@bblimke: I see where you're coming from, and I share some of those concerns. However, these same concerns can be applied to the current Curl::Easy integration. 
As I understand it, the flow of control within curb is 
Curl::Easy http_* -> Curl::Easy#perform -> Curl::Multi.new -> Curl::Multi#perform -> libcurl internals

The ideal place of course, would be to inject the webmock integration within Curl::Multi#perfrom, replacing the libcurl internals. Curl::Multi#perform is a big method though, and does lots of (neccessary) stuff for callback handling, file handle selection, multiplexing and the like. I don't really see a way to inject any code in there without changing curb code. If it were possible, we could throw out about 70% of the existing Curl::Easy webmock integration code.

Currently Curl::Easys integration just replaces the Curl::Easy#perform method and circumvents the Curl::Multi internals (such as callback handling and the like) completely. My reasoning was: If that is currently Good Enough(tm), I can leverage the already existing structure. 

If @taf2 could implement the #getinfo method as a dual to the #setopt method to read out the set CURLOPT_* values, and extract the calls to the libcurl-internals from Curl::Multi.perform, that would make things a lot easier. I could also have a go at it as well, however my available time is currently limited, as is my knowledge of the libcurl api, so it make take a few months until we get something.

As it is, I don't see another way to add the Curl::Multi integration... :( Open to suggestions though.

**Comment by sriedel** [comment]
Wrt separating out a webmock-curlmulti, that would be a way forward. Add a big disclaimer that things may not work as expected. I'll ponder that a bit, however I still see the same issue with Curl::Easy.

**Comment by bblimke** [comment]
The main difference is when working with "real requests", not stubbed requests. These are stored in `RequestRegistry` and are also passed to WebMock request callbacks. I.e VCR project uses that.

When you make a real request with Curl::Easy, then it's going to use Curl::Easy internals to make the request. WebMock will just "listen" and record the request.

With your Curl::Multi wrapper implementation, this would not be the case anymore.

**Comment by sriedel** [comment]
Ah, I see. If that use case is your main concern, I think I can have a look if I can rewrite that part to use the underlying Curl::Multi methods as much as possible. I thought your concerns were of a more general nature :)

**Comment by bblimke** [comment]
Option 1 is used in all other adapter. The decision is made on per call basis.
